---
title: "Sequencing assays"
author:
  - Stefano Mangiola, South Australian immunoGENomics Cancer Institute^[<mangiola.stefano@adelaide.edu.au>], Walter and Eliza Hall Institute^[<mangiola.s at wehi.edu.au>]
output: rmarkdown::html_vignette
# bibliography: "`r file.path(system.file(package='tidySpatialWorkshop2024', 'vignettes'), 'tidyomics.bib')`"
vignette: >
  %\VignetteIndexEntry{Sequencing assays}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = FALSE)

```


::: {.note}
**Exercise 1.1**
:::

```{r, fig.width=7, fig.height=8, eval=FALSE}
# Label
colData(spatial_data)$macro_cluster = reducedDim(spatial_data, "UMAP")[,"UMAP1"] > -2.5

# Verify
scater::plotUMAP(spatial_data, colour_by = "macro_cluster", point_size = 0.2) 

ggspavis::plotVisium(
  spatial_data, 
  annotate = "macro_cluster", 
  highlight = "in_tissue"
)
```


::: {.note}
**Exercise 1.2**
:::

```{r, fig.width=7, fig.height=8, eval=FALSE}

spe_joint <- do.call(cbind, spatial_data_list)


 ggspavis::plotSpots(spe_joint, annotate = sprintf("%s", "clust_M0_lam0.2_k50_res0.7"), pal = pal) +
    facet_wrap(~sample_id) +
    theme(legend.position = "none") +
    labs(title = "BANKSY clusters")

```

```{r, fig.width=7, fig.height=8, eval=FALSE}

spe_joint$has_changed =  !spe_joint$clust_M0_lam0.2_k50_res0.7 == spe_joint$clust_M0_lam0.2_k50_res0.7_smooth

plotSpotQC(
  spe_joint, 
  plot_type = "spot",  
  annotate = "has_changed", 
) + 
  facet_wrap(~sample_id)
```


::: {.note}
**Excercise 1.3**
:::


```{r, fig.width=7, fig.height=8}

res_spatialLIBD = split(data.frame(res$mat), colData(spatial_data_gene_name)$sample_id ) 

lapply(res_spatialLIBD, function(x) plotCorrelationMatrix(as.matrix(x[,-10]))) 

```

::: {.note}
**Excercise 1.4**
:::


```{r, fig.width=7, fig.height=8}

res_spatialLIBD = split(data.frame(res$mat), colData(spatial_data_gene_name)$spatialLIBD ) 

lapply(res_spatialLIBD, function(x) plotCorrelationMatrix(as.matrix(x[,-10]))) 
```

::: {.note}
**Excercise 1.5**
:::


```{r, fig.width=7, fig.height=8}



is_endothelial_leptomeningeal = mat_df$endothelial_cell >0.1 & mat_df$leptomeningeal_cell>0.1 & mat_df$endothelial_cell + mat_df$leptomeningeal_cell > 0.4 

spatial_data$is_endothelial_leptomeningeal = is_endothelial_leptomeningeal

ggspavis::plotSpots(spatial_data, annotate = "is_endothelial_leptomeningeal") +
    facet_wrap(~sample_id) +
  scale_color_manual(values = c("TRUE"= "red", "FALSE" = "grey"))
theme(legend.position = "none") +
  labs(title = "endothelial + leptomeningeal")




is_endothelial_oligodendrocytes = mat_df$endothelial_cell >0.1 & mat_df$oligodendrocyte>0.05 & mat_df$endothelial_cell  + mat_df$oligodendrocyte > 0.4 

spatial_data$is_endothelial_oligodendrocyte = is_endothelial_oligodendrocytes

ggspavis::plotSpots(spatial_data, annotate = "is_endothelial_oligodendrocyte") +
    facet_wrap(~sample_id) +
  scale_color_manual(values = c("TRUE"= "blue", "FALSE" = "grey"))
theme(legend.position = "none") +
  labs(title = "endothelial + oligodendrocyte")

```



::: {.note}
**Exercise 2.1**
:::

```{r}
# Get top variable genes 
genes <- !grepl(pattern = "^Rp[l|s]|Mt", x = rownames(spatial_data))
hvg = scran::modelGeneVar(spatial_data, subset.row = genes, block = spatial_data$sample_id) 
hvg = scran::getTopHVGs(dec, n = 1000)

# Calculate PCA
spatial_data <- 
  spatial_data |> 
  scuttle::logNormCounts() |> 
  scater::runPCA(subset_row = hvg) |> 

  # Calculate UMAP
  scater::runUMAP(dimred = "PCA") |> 

  # Plot
  scater::plotUMAP(colour_by = ".gate")
```


::: {.note}
**Exercise 2.2**
:::


```{r}
rowData(spatial_data) |> 
  as_tibble() |> 
  filter( gene_name == "PECAM1")

spatial_data |> 

  # Join the feature
  join_features("ENSG00000261371", shape="wide") |> 

  # Calculate the quantile
  mutate(my_quantile = quantile(ENSG00000261371, 0.75)) |> 

  # Label the pixels
  mutate(PECAM1_positive = ENSG00000261371 > my_quantile) |> 

  # Plot
  ggspavis::plotSpots(annotate = "PECAM1_positive") +
  facet_wrap(~sample_id) 

```


::: {.note}
**Excercise 2.3**
:::

```{r}
library(tidySummarizedExperiment)
library(tidybulk)
  
differential_analysis = 
  spatial_data |> 
  mutate(
    dead = 
      
      # Stringent threshold
      subsets_mito_percent > 20 |
      sum < 700 |
      detected < 500
  ) |> 
  aggregate_cells(c(sample_id, spatialLIBD, dead)) |> 
  keep_abundant(factor_of_interest = c(dead)) |> 
  nest(data = - spatialLIBD) |> 
  
  # filter regions having both alive and dead cells
  filter(  map_int(data, ~ .x |> distinct(sample_id, dead) |> nrow() ) == 6 ) |>
  mutate(data = map(
    data,
    test_differential_abundance,
    ~ dead + sample_id,
    method = "edgeR_quasi_likelihood", 
    test_above_log2_fold_change = log(2)
  )) 

differential_analysis |> 
  mutate(data = map(data, pivot_transcript)) |> 
  unnest(data) |> 
  filter(FDR<0.05) 
```


::: {.note}
**Excercise 2.4**
:::

```{r}
rownames(spatial_data) = rowData(spatial_data)$gene_name

marker_genes_of_amyloid_plaques = c("APP", "PSEN1", "PSEN2", "CLU", "APOE", "CD68", "ITGAM", "AIF1")

spatial_data |> 

# Join the features
  join_features(marker_genes_of_amyloid_plaques, shape = "wide") |> 

  # Rescaling
  mutate(across(any_of(marker_genes_of_amyloid_plaques), scales::rescale)) |> 

# Summarising signature
  mutate(amyloid_plaques_signature  = APP + PSEN1 + PSEN2 + CLU + APOE + CD68 + ITGAM + AIF1) |> 

# Plotting
  ggspavis::plotSpots(
    annotate = "amyloid_plaques_signature"
  ) + 
  facet_wrap(~sample_id)


```